### PR TITLE
Implement native Go TOML parser

### PR DESCRIPTION
**Pull request by azinneera**
## Summary by CodeRabbit

* **New Features**
  * Ship a native TOML parser with broad syntax support and a public To() API for mapping into structs; improved diagnostics with precise locations and error recovery.

* **Tests**
  * Expanded test suite and many new TOML fixtures covering syntax, edge cases, and error-recovery scenarios.

* **Chores**
  * Removed external TOML dependency in favor of the native implementation.

### CODE REVIEW

File: common/tomlparser/internal/ast/inline_table_node.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		case *KeyValueNode:
+			m[k] = tv.NativeValue()
+		case *TableNode:
+			m[k] = tv.ToMap()
+		default:
+			m[k] = nil
+		}
+	}
+	return m
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/inline_table_node.go:74]: `TableArrayNode` entries come from `[[...]]` headers, which  
  are top-level constructs — the TOML spec forbids them inside inline tables and the parser never routes them there. This case is unreachable. So no changes needed.

File: common/tomlparser/internal/lexer/input_reader.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+// PeekAt returns the rune at position pos+k (0-indexed offset from current pos).
+// Returns charEOF if out of bounds.
+func (r *InputReader) PeekAt(k int) rune {
+	idx := r.pos + k
+	if idx >= len(r.runes) {
+		return charEOF
+	}
+	return r.runes[idx]
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/lexer/input_reader.go:62]: There is no code path that supplies a negative `k` to `PeekAt`. So no change needed.

File: common/tomlparser/internal/parser/token_reader.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		r.size++
+	}
+	if k > r.size {
+		k = r.size
+	}
+	if k <= 0 {
+		return lexer.Token{Kind: lexer.TokenEOF}
+	}
+	return r.buf[(r.start+k-1)%tokenBufferCap]
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/token_reader.go]: The `tokenBufferCap` of 20 provides enough margin. The usual dotted keys are not expected to go beyond this at all.

File: common/tomlparser/internal/parser/toml_parser.go [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+
+func (p *Parser) parseNumericValue() ast.ValueNode {
+	sign := ""
+	startTok := p.reader.Peek()
+
+	if startTok.Kind == lexer.TokenPlus || startTok.Kind == lexer.TokenMinus {
+		p.reader.Read()
+		if startTok.Kind == lexer.TokenMinus {
+			sign = "-"
+		}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go:542]: This will be addressed in the next phase

File: common/tomlparser/toml_test.go [hunk 5]
```diff
@@ -421,6 +421,213 @@ title = "Duplicate"
...
+		t.Fatalf("failed to parse comprehensive-ballerina.toml: %v", err)
+	}
+	return doc
+}
+
+func TestBallerinaTomlParsesWithoutErrors(t *testing.T) {
+	readComprehensiveBallerinaToml(t)
+}
+
+func TestBallerinaToml_PackageScalarFields(t *testing.T) {
```
Comment (warunalakshitha) [inline common/tomlparser/toml_test.go:490]: n some cases we use camelCase for function names, and in some cases we use underscores. Shall we have consistent naming?

File: common/tomlparser/internal/ast/float_value_node.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (warunalakshitha) [inline common/tomlparser/internal/ast/float_value_node.go:1]: file name float_value_node.go

File: common/tomlparser/internal/ast/key_value_node.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+// NewKeyValueNode creates a single-key (non-dotted) KeyValueNode.
+func NewKeyValueNode(key string, value ValueNode, loc Location) *KeyValueNode {
+	return &KeyValueNode{keys: []string{key}, value: value, loc: loc}
+}
+
+// NewKeyValueNodeWithPath creates a dotted-key KeyValueNode.
+// Used by the parser when it encounters a.b.c = val.
+func NewKeyValueNodeWithPath(keys []string, value ValueNode, loc Location) *KeyValueNode {
+	return &KeyValueNode{keys: keys, value: value, loc: loc}
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/key_value_node.go:39]: This is handled at the caller

File: common/tomlparser/internal/ast/key_value_node.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+func (n *KeyValueNode) Kind() TomlType   { return TypeKeyValue }
+func (n *KeyValueNode) Loc() Location    { return n.loc }
+func (n *KeyValueNode) KeyName() string  { return n.keys[len(n.keys)-1] }
+func (n *KeyValueNode) Keys() []string   { return n.keys }
+func (n *KeyValueNode) Value() ValueNode { return n.value }
+func (n *KeyValueNode) NativeValue() any { return n.value.NativeValue() }
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/key_value_node.go:46]: nil values are checked in toml_parser.go in `parseKeyValue`. Additional nil check here is not required.

File: common/tomlparser/internal/parser/toml_parser.go [hunk 9]
```diff
@@ -0,0 +1,854 @@
...
+		table.AddEntry(kv.KeyName(), kv)
+		return
+	}
+	// Dotted key inside inline table — create sub-tables.
+	// We model this as a TableNode stored as a KeyValueNode with a TableNode value placeholder.
+	// For simplicity in P1, just use the last key segment.
+	leaf := ast.NewKeyValueNode(keys[len(keys)-1], kv.Value(), kv.Loc())
+	table.AddEntry(leaf.KeyName(), leaf)
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go:844]: This will be addressed in the next phase

File: common/tomlparser/toml.go [hunk 10]
```diff
@@ -97,22 +95,21 @@ func ReadStreamWithSchema(reader io.Reader, schema Schema) (*Toml, error) {
...
 
-	if err != nil {
-		t.diagnostics = append(t.diagnostics, parseErrorDiagnostic(err))
+	if len(parseErrors) > 0 {
+		return t, fmt.Errorf("TOML parse error: %s", parseErrors[0].Message)
 	}
-
-	return t, err
+	return t, nil
 }
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml.go:113]: root is created with `ast.NewTableNode(...)` before any parsing begins and is never reassigned. Therefore, there is no chance of returning a `nil` here

File: common/tomlparser/toml_test.go [hunk 11]
```diff
@@ -421,6 +421,213 @@ title = "Duplicate"
...
+	})
+	t.Run("modules[1]", func(t *testing.T) {
+		name, _ := modules[1].GetString("name")
+		if name != "my_service.db" {
+			t.Errorf("got %q, want my_service.db", name)
+		}
+		export, _ := modules[1].GetBool("export")
+		if export {
+			t.Error("modules[1].export should be false")
+		}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_test.go:604]: Fixed in 0899b5169d1077c00483595cb965b41dda238b3f

File: common/tomlparser/internal/ast/float_value_node.go (hunk 6 above)
Comment (azinneera) [reply common/tomlparser/internal/ast/float_value_node.go:1]: Fixed in 0899b5169d1077c00483595cb965b41dda238b3f

File: common/tomlparser/internal/ast/inline_table_node.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	for k, v := range n.entries {
+		switch tv := v.(type) {
+		case *KeyValueNode:
+			m[k] = tv.NativeValue()
+		case *TableNode:
+			m[k] = tv.ToMap()
+		default:
+			m[k] = nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/inline_table_node.go:72]: Fixed in 0899b5169d1077c00483595cb965b41dda238b3f

File: common/tomlparser/internal/ast/table_node.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+	if !ok || !existingTable.generated {
+		return
+	}
+	// Merge existing entries into the replacement table.
+	for k, v := range existingTable.entries {
+		if _, exists := newTable.entries[k]; !exists {
+			newTable.order = append(newTable.order, k)
+		}
+		newTable.entries[k] = v
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/table_node.go]: Fixed in ce84da92a0d314bc733032c21b25a82c4e08e51e

File: common/tomlparser/toml_test.go (hunk 5 above)
Comment (azinneera) [reply common/tomlparser/toml_test.go:490]: Some tests are added with the `_` to specify a group. These tests are to verify the Ballerinatoml structure in specific. Hence the naming pattern. AFAIK, this is an accepted convention in Go

File: common/tomlparser/internal/lexer/toml_lexer.go [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+}
+
+// SwitchMode replaces the top of the stack with m (pop + push).
+func (l *Lexer) SwitchMode(m ParserMode) {
+	if len(l.stack) > 0 {
+		l.stack = l.stack[:len(l.stack)-1]
+	}
+	l.stack = append(l.stack, m)
+	l.mode = m
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/lexer/toml_lexer.go]: The proposed fix is semantically identical to the current code. No code changes necessary here

File: common/tomlparser/toml_test.go [hunk 15]
```diff
@@ -441,8 +654,62 @@ key = "value"
...
+// overwritten by the generated (implicit) entry during the merge.
+func TestGeneratedTableMergeConflict(t *testing.T) {
+	doc, _ := Read(os.DirFS("testdata"), "generated-table-merge-conflict.toml")
+	if doc == nil {
+		t.Fatal("expected partial document even on error")
+	}
+	if got, ok := doc.GetString("fruit.color"); !ok || got != "explicit" {
+		t.Errorf("fruit.color = %q, want \"explicit\" (explicit must not be overwritten by generated)", got)
+	}
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_test.go]: The test has a duplicate key which is intentional and it expects a partial document even on error

File: common/tomlparser/toml_java_compat_test.go [hunk 16]
```diff
@@ -0,0 +1,465 @@
...
+	expectStr(t, doc, "key", "hello")
+}
+
+// TestJava_NegMultiErrors — neg-multi-errors.toml: three error lines
+// (key"value", ="value", key3 =) interspersed with valid lines; valid keys
+// key1 and key2 must survive error recovery.
+func TestJava_NegMultiErrors(t *testing.T) {
+	doc := mustLoadWithError(t, "neg-multi-errors.toml")
+	expectDiagCount(t, doc, 1)
+	expectInt(t, doc, "key1", 12)
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_java_compat_test.go:438]: Fixed in 3fdc1c8d921604cd8dc59a65030a9de179d1313d

File: common/tomlparser/toml_java_compat_test.go [hunk 17]
```diff
@@ -0,0 +1,465 @@
...
+
+	integers, ok := doc.GetArray("integers")
+	if !ok {
+		t.Fatal("integers array not found")
+	}
+	if len(integers) != 3 {
+		t.Errorf("integers: len=%d, want 3", len(integers))
+	}
+	if v, _ := integers[0].(int64); v != 1 {
+		t.Errorf("integers[0] = %v, want 1", integers[0])
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_java_compat_test.go:191]: Fixed in 3fdc1c8d921604cd8dc59a65030a9de179d1313d

File: common/tomlparser/internal/lexer/toml_lexer.go [hunk 18]
```diff
@@ -0,0 +1,714 @@
...
+
+// processBinaryLiteral reads a 0b... binary integer.
+func (l *Lexer) processBinaryLiteral(line, col int) Token {
+	l.reader.Advance() // consume 'b' or 'B'
+	for isBinaryDigit(l.reader.Peek()) {
+		l.reader.Advance()
+	}
+	lexeme := l.reader.GetMarkedChars()
+	return Token{Kind: TokenBinaryInt, Value: lexeme, Line: line, Column: col}
+}
```
Comment (azinneera) [inline common/tomlparser/internal/lexer/toml_lexer.go:618]: Fixed in 3fdc1c8d921604cd8dc59a65030a9de179d1313d

File: common/tomlparser/internal/lexer/toml_lexer.go [hunk 19]
```diff
@@ -0,0 +1,734 @@
...
+		digitSeen = true
+		l.reader.Advance()
+	}
+	lexeme := l.reader.GetMarkedChars()
+	if !digitSeen {
+		l.addError("missing digit after base prefix", line, col)
+		return Token{Kind: TokenInvalid, Value: lexeme, Line: line, Column: col}
+	}
+	return Token{Kind: TokenBinaryInt, Value: lexeme, Line: line, Column: col}
+}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/lexer/toml_lexer.go:618]: False positive. Added a test case to cover the scenario and it works as expected.

File: common/tomlparser/toml.go [hunk 20]
```diff
@@ -261,10 +255,139 @@ func (t *Toml) ToMap() map[string]any {
...
+		// Also store by lowercased field name for case-insensitive matching,
+		// but only if that slot is not already claimed by a tag from an earlier
+		// field — otherwise the name-based fallback would silently overwrite a
+		// deliberate tag-based mapping.
+		if lowerName := strings.ToLower(f.Name); lowerName != tomlKey {
+			if _, exists := lookup[lowerName]; !exists {
+				lookup[lowerName] = fieldDesc{jsonName: jsonName, fieldType: f.Type}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml.go:330]: Fixed in [92e5118](https://github.com/ballerina-platform/ballerina-lang-go/pull/218/commits/92e511865742cabbdcfeb69e6582fe9f251b1b36)

File: common/tomlparser/toml_test.go [hunk 21]
```diff
@@ -441,8 +682,81 @@ key = "value"
...
+// boundary so that the next valid key-value pair is still parsed.
+// testdata/error-recovery.toml contains:
+//
+//	bad_key        ← missing '=', triggers skipToRecovery
+//	key = "value"  ← must survive recovery
+//
+// TestGeneratedTableMerge — when a dotted key (e.g. fruit.color = "yellow")
+// implicitly creates a table and a later [fruit] section explicitly defines it,
+// the entries from the generated table must be carried into the explicit one.
+func TestGeneratedTableMerge(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_test.go]: Fixed in [92e5118](https://github.com/ballerina-platform/ballerina-lang-go/pull/218/commits/92e511865742cabbdcfeb69e6582fe9f251b1b36)

File: common/tomlparser/internal/lexer/toml_lexer.go [hunk 22]
```diff
@@ -0,0 +1,734 @@
...
+	if len(l.stack) == 0 {
+		l.mode = ModeDefault
+		return
+	}
+	l.mode = l.stack[len(l.stack)-1]
+	l.stack = l.stack[:len(l.stack)-1]
+}
+
+// CurrentMode returns the active parser mode.
+func (l *Lexer) CurrentMode() ParserMode {
```
Comment (warunalakshitha) [inline common/tomlparser/internal/lexer/toml_lexer.go]: any usage of CurrentMode?

File: common/tomlparser/internal/parser/toml_parser.go [hunk 23]
```diff
@@ -0,0 +1,830 @@
...
+			default:
+				p.addErrorAt(
+					fmt.Sprintf("key %q is not a table", seg),
+					root.Loc().StartLine, root.Loc().StartColumn)
+				return current
+			}
+		} else {
+			newTable := ast.NewGeneratedTableNode(seg, root.Loc())
+			current.AddEntry(seg, newTable)
+			current = newTable
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go]: Fixed in 5dba57a

File: common/tomlparser/internal/parser/toml_parser.go [hunk 24]
```diff
@@ -0,0 +1,830 @@
...
+	// and let addChildKeyValueToTable handle the dotted path.
+	return ast.NewKeyValueNodeWithPath(keys, val, loc)
+}
+
+// parseKeyList reads one or more dot-separated key segments.
+// Returns a slice of key strings (never nil; may be empty on error).
+func (p *Parser) parseKeyList() []string {
+	first := p.parseSingleKeySegment()
+	if first == "" {
+		return nil
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go:351]: Fixed in 5dba57a

File: common/tomlparser/internal/parser/toml_parser.go [hunk 25]
```diff
@@ -0,0 +1,830 @@
...
+	if val == nil {
+		p.skipToRecovery()
+		return nil
+	}
+
+	// Consume trailing newline (if present) — not inside inline table.
+	tok := p.reader.peek()
+	if tok.Kind == lexer.TokenNewline {
+		p.reader.read()
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go:327]: Fixed in 5dba57a

File: common/tomlparser/internal/parser/toml_parser.go [hunk 26]
```diff
@@ -0,0 +1,830 @@
...
+		}
+		if tok.Kind == lexer.TokenNewline {
+			p.reader.read()
+			continue
+		}
+		kv := p.parseKeyValue()
+		if kv != nil {
+			p.addChildKeyValueToTable(tableNode, kv)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/parser/toml_parser.go:215]: Fixed in 5dba57a

File: common/tomlparser/internal/ast/table_node.go [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+	// Merge generated entries into the replacement table.
+	// Only carry over keys that the explicit table has not already defined;
+	// an explicit entry must never be overwritten by a generated (implicit) one.
+	for k, v := range existingTable.entries {
+		if _, exists := newTable.entries[k]; !exists {
+			newTable.order = append(newTable.order, k)
+			newTable.entries[k] = v
+		}
+	}
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/internal/ast/table_node.go:84]: Fixed in 5dba57a

File: common/tomlparser/toml_test.go [hunk 28]
```diff
@@ -441,8 +705,107 @@ key = "value"
...
+	if err == nil {
+		t.Fatal("expected a parse error for the bad_key line, got nil")
+	}
+	if tomlDoc == nil {
+		t.Fatal("expected partial Toml even on parse error")
+	}
+	_, ok := tomlDoc.Get("key")
+	if !ok {
+		t.Error("key = \"value\" was swallowed by overshoot recovery; " +
+			"skipToRecovery must stop at TokenNewline")
```
[the remark this answers is not in the document]
Comment (azinneera) [reply common/tomlparser/toml_test.go]: Fixed in 5dba57a

File: common/tomlparser/internal/lexer/toml_lexer.go (hunk 22 above)
Comment (azinneera) [reply common/tomlparser/internal/lexer/toml_lexer.go]: Removed in 5dba57a
